File: bindings/rust/examples/model_store_demo.rs
use sop::{Context, Database, DatabaseOptions};
use std::fs;
use std::path::Path;
use std::str;

fn main() {
    println!("\n--- Running Model Store Example ---");
    println!("Scenario: Storing Large AI Models (BLOBs)");

    let ctx = Context::new();
    let db_path = "sop_data_model";
    if Path::new(db_path).exists() {
        fs::remove_dir_all(db_path).unwrap();
    }

    let db = Database::new(&ctx, DatabaseOptions {
        stores_folders: Some(vec![db_path.to_string()]),
        ..Default::default()
    }).unwrap();

    let trans = db.begin_transaction(&ctx).unwrap();
    
    // OpenModelStore creates a new store if it doesn't exist
    let model_store = db.open_model_store(&ctx, "llm_weights", &trans).unwrap();

    // 1. Save a "Model" (simulated large binary data)
    let category = "llm";
    let model_name = "gpt-mini-v1";
    let weights = "...simulated large binary model weights...".as_bytes().to_vec();
    
    println!("Saving model '{}' in category '{}' ({} bytes)...", model_name, category, weights.len());
    model_store.save(&ctx, category, model_name, weights.clone()).unwrap();

    // 2. Load the Model
    println!("Loading model back...");
    match model_store.load(&ctx, category, model_name) {
        Ok(loaded_weights) => {
            let content = str::from_utf8(&loaded_weights).unwrap();
            println!("Loaded content: {}", content);
        },
        Err(e) => println!("Error loading model: {}", e),
    }

    // 3. Delete
    println!("Deleting model...");
    model_store.delete(&ctx, category, model_name).unwrap();

    trans.commit(&ctx).unwrap();
    
    println!("--- End of Model Store Demo ---");
}

File: bindings/python/sop/examples/modelstore_demo.py
import os
import shutil
import sys
from dataclasses import dataclass

# Add the parent directory to sys.path to import sop
# sys.path.append(os.path.join(os.path.dirname(__file__), '..'))

from sop import Context
from sop.ai import Database, DatabaseType
from sop.database import DatabaseOptions

@dataclass
class MyModel:
    name: str
    version: str
    parameters: dict
    accuracy: float

def main():
    db_path = "model_store_demo_db"
    
    # Clean up previous run
    if os.path.exists(db_path):
        shutil.rmtree(db_path)

    print(f"Initializing SOP Database at '{db_path}'...")
    ctx = Context()
    # Initialize the Database (Unified Mode)
    db = Database(DatabaseOptions(stores_folders=[db_path], type=DatabaseType.Standalone))

    # Start Transaction
    with db.begin_transaction(ctx) as trans:
        print("Transaction Started.")

        # Open a Model Store
        print("Opening Model Store 'experiment_1'...")
        store = db.open_model_store(ctx, trans, "experiment_1")

        # Create a sample model object
        model = MyModel(
            name="gpt-4-mini",
            version="1.0.0",
            parameters={"layers": 12, "heads": 8},
            accuracy=0.95
        )

        # Save the model
        print(f"Saving model '{model.name}' to category 'llm'...")
        store.save(ctx, category="llm", name=model.name, model=model)

        # List models in the category
        print("Listing models in category 'llm'...")
        models = store.list(ctx, category="llm")
        print(f"Found models: {models}")

        # Load the model back
        print(f"Loading model '{model.name}'...")
        loaded_data = store.get(ctx, category="llm", name=model.name)
        print(f"Loaded data: {loaded_data}")

        # Verify data
        if loaded_data['name'] == model.name and loaded_data['accuracy'] == model.accuracy:
            print("Verification Successful: Model data matches.")
        else:
            print("Verification Failed!")

        # Delete the model
        print(f"Deleting model '{model.name}'...")
        store.delete(ctx, category="llm", name=model.name)

        # Verify deletion
        models_after = store.list(ctx, category="llm")
        print(f"Models after deletion: {models_after}")

        if not models_after:
            print("Deletion Successful.")
        
        print("Committing Transaction...")

    # Clean up
    if os.path.exists(db_path):
        shutil.rmtree(db_path)
    print("Demo completed successfully.")

if __name__ == "__main__":
    main()
